pandas/core/frame.py

```diff
@@ -2018,7 +2018,14 @@ def _getitem_array(self, key):
             indexer = key.nonzero()[0]
             return self.take(indexer, axis=0, convert=False)
         else:
-            indexer = self.ix._convert_to_indexer(key, axis=1)
+            try:
+                indexer = self.ix._convert_to_indexer(key, axis=1)
+            except KeyError:
+                if (hasattr(self, 'index') and
+                        any(item in self.index.names for item in key)):
+                    return self._getitem_array_with_index_name(key)
```

Comment:
this is still way too messy. This should all be encompassed in a single function that is invoked on a ``KeyError``